### ISSUE TITLE
Add doc note about conversion of a value nested in a promise

### ISSUE THREAD

**Issue by cdoublev** [jsdom/webidl2js#257]
> However, note that apart from Web IDL container return values, this impl-back-to-wrapper conversion doesn't work in a "deep" fashion. That is, if you directly return an impl, or return an array or object containing impls from a sequence<>, FrozenArray<>, or record<>-returning operation, the conversion will work.

Because a `<Promise>` is defined in [Web IDL](https://webidl.spec.whatwg.org/#idl-promise) but *there is no way to represent a promise value in IDL*, a return value defined in Web IDL and wrapped in a promise will not be converted when the promise resolves, therefore I suggest the following modification:

```diff
- However, note that apart from Web IDL container return values, this impl-back-to-wrapper conversion doesn't work in a "deep" fashion.
+ However, note that apart from container return values that can be defined in Web IDL, this impl-back-to-wrapper conversion doesn't work in a "deep" fashion.
```

**Comment by domenic** [comment]
I'm not sure what you mean by "there is no way to represent a promise value in IDL", or "a return value defined with Web IDL and wrapped in a promise will not be converted when the promise resolve". Could you give a code example?

**Comment by cdoublev** [comment]
This is the description used in the Web IDL specification. I naively expected that an instance of a class defined with IDL and wrapped in a promise, as a return value from a class method also defined with Web IDL, could be returned as an instance of its wrapper class, due to what I remembered from the related section in the `webidl2js` documentation, and from rereading it again after I realized that it is not converted.

My suggestion is just an attempt to clarify that despite the presence of the `Promise` type in the web IDL specification, an instance wrapped in a promise can not be converted to an instance of the container class.

**Comment by domenic** [comment]
> This is the description used in the Web IDL specification.

This means there's no syntax for producing promise values, like `1` is syntax for producing a numeric value.

>  I naively expected

Sorry, this is quite hard to follow for me. Could you give a code example of what you wrote in the impl class, what you expected to happen, and what happened instead?

**Comment by cdoublev** [comment]
It's my fault, I'm in a rush to finish this asap. For `CSSStyleSheetImpl.replace()`:

```js
replace(text) {
    /* not meaningfull */
    return Promise.resolve().then(() => {
      /* not meaningfull */
      return this
      /* Workaround: return wrapperForImpl(this) */
  })
}
```

Then in my test file:

```js
const styleSheetRef = await styleSheet.replace(/* not meaningfull */)
expect(CSSStyleSheet.is(styleSheetRef))
```

Before using the above workaround (`wrapperForImpl()`), I got back a `CSSStyleSheetImpl` instead of `CSSStyleSheet`.

**Comment by domenic** [comment]
Got it, yes, I can see how the docs aren't sufficiently clear here.

In fact, looking further it seems like they might just be wrong? I don't see any code for us to handle the `sequence<>`, `FrozenArray<>`, or `record<>` cases that the current README mentions. (We definitely don't handle the `Promise<>` case, except to convert any thrown exceptions into rejections.)

The ideal fix for this would be finishing up https://github.com/jsdom/webidl2js/pull/108.

In the meantime, the way we work around this is to write slightly-more-awkward impl classes, which use wrapperForImpl. E.g. [here](https://github.com/jsdom/jsdom/blob/04f6c13f4a4d387c7fc979b8f62c6f68d8a0c639/lib/jsdom/living/nodes/HTMLSlotElement-impl.js#L42-L49) is an example of manually wrapping a `sequence<>`. I think that would work for your case?
